verification.go

```diff
@@ -0,0 +1,154 @@
+// correctly and according to the rules of BIP-0158 to contain both the output's
+// pk scripts as well as the pk scripts the inputs are spending.
+func VerifyBasicBlockFilter(filter *gcs.Filter, block *btcutil.Block) (int,
+	error) {
+
+	var (
+		opReturnMatches int
+		key             = builder.DeriveKey(block.Hash())
+	)
+	for idx, tx := range block.Transactions() {
+		// Skip coinbase transaction.
+		if idx == 0 {
+			continue
+		}
+
+		// Check outputs first.
+		for outIdx, txOut := range tx.MsgTx().TxOut {
+			switch {
+			// If the script itself is blank, then we'll skip this
+			// as it doesn't contain any useful information.
+			case len(txOut.PkScript) == 0:
+				continue
+
+			// We'll also skip any OP_RETURN scripts as well since
+			// we don't index these in order to avoid a circular
+			// dependency.
+			case txOut.PkScript[0] == txscript.OP_RETURN:
+				// Previous versions of the filters did include
+				// OP_RETURNs. To be able disconnect bad peers
+				// still serving these old filters we attempt to
+				// check if there's an unexpected match. Since
+				// there might be false positives, an OP_RETURN
+				// can still match filters not including them.
+				// Therefore, we count the number of such
+				// unexpected matches for each peer, such that
+				// we can ban peers matching more than the rest.
+				match, err := filter.Match(key, txOut.PkScript)
+				if err != nil {
+					// Mark peer bad if we cannot match on
+					// its filter.
+					return 0, fmt.Errorf("error "+
+						"validating block %v outpoint "+
+						"%v:%d script %x: %v",
+						block.Hash(), tx.Hash(), outIdx,
+						txOut.PkScript, err)
+				}
+
+				// If it matches on the OP_RETURN output, we
+				// increase the op return counter.
+				if match {
+					opReturnMatches++
+				}
+
+				continue
+			}
+
+			// This is a "normal" script where we definitely expect
+			// a match.
+			match, err := filter.Match(key, txOut.PkScript)
+			if err != nil {
+				return 0, fmt.Errorf("error validating block "+
+					"%v outpoint %v:%d script %x: %v",
+					block.Hash(), tx.Hash(), outIdx,
+					txOut.PkScript, err)
+			}
+
+			if !match {
+				return 0, fmt.Errorf("filter for block %v is "+
+					"invalid, outpoint %v:%d script %x "+
+					"wasn't matched by filter",
+					block.Hash(), tx.Hash(), outIdx,
+					txOut.PkScript)
+			}
+		}
+
+		// Now we can go through all inputs and check that the filter
+		// also included any pk scripts of the outputs being _spent_.
+		// We can do this for witness items since the witness always
+		// contains the full script as the last element on the stack.
+		for inIdx, in := range tx.MsgTx().TxIn {
+			// There are too many edge cases to cover for non-
+			// witness scripts. And in LN land we're interested in
+			// witness spends only anyway. Therefore let's skip any
+			// input that has no witness.
+			//
+			// TODO(guggero): Add all those edge cases to
+			// ComputePkScript?
+			if len(in.Witness) == 0 {
```

Comment:
`ComputePkScript` supports `P2PKH` as well, so we could opt to verify those as well for users still using those addresses.

Reply:
You mean run `ComputePkScript` even if there is no witness and see if the returned script is P2PKH and only then verify it? I'm a big afraid that we'd run into another edge case here. If I look at the `computeNonWitnessPkScript` function, it looks like we could be very unlucky and if a P2PK script has a 0x02 or 0x03 in the wrong place of the signature, this would actually match and we'd fail the whole block for it.
I'd have a better feeling if we added more cases to the non-witness script computation first before we fail the verification here.

Reply:
Ah I see, good points. Probably not worth the hassle anyway.